### ISSUE TITLE
Front-end config should be defined on handlers?

### ISSUE THREAD

**Issue by jonom** [colymba/GridFieldBulkEditingTools#109]
At the moment the default front-end config settings that apply to a `GridFieldBulkActionHandler` are set in `GridFieldBulkManager::__construct()`. Here are the issues I see with that:
1. If I subclass `GridFieldBulkActionHandler` I can't define default front-end config settings for my new handler because I can't modify the `__construct` function on `GridFieldBulkManager`. I would need to pass in all the necessary settings whenever I add this action to a `GridFieldBulkManager` instance.
2. It doesn't seem advantageous to be able to be able to override config settings for a handler in the first place. `isAjax` and `isDestructive` are surely tied to the behaviour of the handler in question and only have one valid value per handler so shouldn't me modified. `icon` could me modified I guess if a developer doesn't like the default icon specified but I think this is overkill. 'Convention over configuration' could be applied here to simplify things.

I think the front-end config settings should be defined on each handler, and possibly the ability to override these settings should be removed. I just seems like an unnecessary complication at the moment... unless I'm missing something?

Let me know if you'd consider accepting a PR for this change and I can work on it. Thanks for the module!

**Comment by colymba** [comment]
My first thought... I agree, let's do it!

Couple of thoughts, yes let's put all the config on the Handler and have an easy way to retrieve it with something like `get_config` so it's easy to build the html. Or something else...

We should then be able to simplify the `addBulkAction` and `removeBulkAction` by just passing the handler's class name? And just hold a list of handler's class names on the `bulkManager` instead of the full on config lists.

Is it worth creating an `Interface` too? So we can define what each handler must have?

**Comment by jonom** [comment]
Maybe it could be like:
```php
addBulkAction(String $handlerClassName, Array $settings=null)
```
This would allow for simple use with e.g. `$manager->addBulkAction('GridFieldBulkActionUnLinkHandler')`, but also allows each handler to introduce unique setting options in an unconstrained way. The `$settings` could be merged with a `$default_settings` static set on the handler so you could pass through all the applicable settings or just one or two and use defaults for the rest. (I'm sure this is some kind of a pattern but couldn't find much about it with a spot of googling)

Those settings could potentially be used when setting up the UI on the gridfield (that's where an `Interface` would help), and also passed through to the handler action. Example:
```php
$manager->addBulkAction(
	'GridFieldBulkActionMoveToFolderHandler',
	array(
		'targetFolderID' => 5,
		'label' => 'Move up one level'
	)
);
```

If above was possible you might want to add the same type of action more than once with different settings so we'd want to make sure that was allowed.

**Comment by colymba** [comment]
This all sounds good. :+1: 

One thing I'm thinking about, if having the ability to add the same bulkAction more than once with different setting could maybe be useful. It would make removing bulkAction more *difficult*, at least we will need a reference name we could use.

So maybe we'll need another optional param in `addBulkAction`, a reference name (in your example it could be "moveToFolderOneTop") that can be used to remove the action. While using the class name to remove an action still works, it would remove ALL actions with that class name.

So we would have:

`addBulkAction(String $handlerClassName, Array $settings=null, String $reference=null)`

````
$manager->addBulkAction(
    'GridFieldBulkActionMoveToFolderHandler',
    array(
        'targetFolderID' => 5,
        'label' => 'Move up one level'
    ),
   'moveToFolderOneTop'
);
````

`removeBulkAction(String $classOrReference)`

`$manager->removeBulkAction('GridFieldBulkActionMoveToFolderHandler');`
`$manager->removeBulkAction('moveToFolderOneTop');`

**Comment by jonom** [comment]
Ah good point! I wonder though, since this would be an optional setting with a default value if this should just be one of the settings. So you could do:
```php
$manager->addBulkAction(
	'GridFieldBulkActionMoveToFolderHandler',
	array(
		'targetFolderID' => 5,
		'label' => 'Move up one level',
		'reference' => 'moveToFolderOneTop'
	)
);
```
In practical terms though I feel like the `removeBulkAction` method probably isn't/won't be used very often, except to remove a single field from a `GridFieldBulkManager` created with default settings. If people have manually added actions they're unlikely to want to remove them I think. So I wonder if we consider real world rather than theoretical use, if it is necessary to introduce the `reference` property? Maybe it's not so bad if `removeBulkAction` removes all the handlers of a particular class because it might be unlikely to affect many people and only cause a minor annoyance when it does. It could be a feature that is added later if it turns out that people really are being affected by this limitation?

**Comment by colymba** [comment]
Alright, let's not bother with a reference at this point. :)

Are you working on a PR?

**Comment by jonom** [comment]
I haven't started on the work yet but I can try to get a PR through in the next week or two unless you'd rather take it on?

**Comment by colymba** [comment]
I am more than happy for you to take it on. Not sure if I could do it quicker than you at the minute.
Looking forward to the PR :)

**Comment by jonom** [comment]
Okay! I'll get in touch when I have something to show.

**Comment by colymba** [comment]
this has now been implemented in some form in the 3.0 branch 5711998ef2ed3ecd51111ac41b17a80436cd1a21
